cirq-core/cirq/testing/consistent_unitary_test.py

```diff
@@ -0,0 +1,72 @@
+# Copyright 2023 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import cirq
+
+import pytest
+import numpy as np
+
+
+class InconsistentGate(cirq.Gate):
+    def _num_qubits_(self) -> int:
+        return 1
+
+    def _unitary_(self) -> np.ndarray:
+        return np.eye(2, dtype=np.complex128)
+
+    def _decompose_with_context_(self, qubits, *, context):
+        (q,) = context.qubit_manager.qalloc(1)
+        yield cirq.X(q)
+        yield cirq.CNOT(q, qubits[0])
+
+
+class FailsOnDecompostion(cirq.Gate):
```

Comment:
Can you add a gate that incorrectly borrows a qubit? So maybe a 

```python
class CZ:
    def _num_qubits_(self):
        return 2

    def _decompose_with_context_(self, qubits, *, context):
        anc = context.qubit_manager.qborrow(1)
        yield cirq.CCNOT(*qubits, *anc)
        yield cirq.Z(*anc)
        yield cirq.CCNOT(*qubits, *anc)
```

This is an example where `context.qubit_manager.qalloc`  should be correct but `context.qubit_manager.qborrow` would result in assertion failure.